CloudSchedule

pennam opened the pull request:
Opening to start integration testing with cloud side.

/cc @eclipse1985

TODO:
- ~~Add property tests~~
- ~~Use const instead of numbers~~
- ~~Check CloudSchedule isActive() method~~
- ~~Timezone -> see https://github.com/pennam/ArduinoIoTCloud/tree/cloudschedule_timezone~~

aentinger reviewed (changes requested):
Hi @pennam :wave: :coffee: Can I ask you to add a example sketch show-casing how to actually use that property? From looking at the source code I don't really get it :confused:

pennam commented:
@aentinger sure!

pennam commented:
Hi @aentinger, i've added a simple sketch just to show how the cloud schedule property can be used.
I must admit that configuring a schedule from a sketch is really ugly, but the configuration part should be handled through the Arduino Create scheduler widget.

To define a schedule 4 field needs to be provided:
`frm`: uint32_t unix timestamo of schedule starting date and time
`to `: uint32_t unix timestamp of schedule end date and time. 0 if schedule has no end.
`len`: uint32_t duration of scheduled event
`msk`: uint32_t bitfield schedule configuration value.

` | 2bits scheduleUnit | 4bits scheduleType | 26 bits schedule Mask |`

```
enum class ScheduleUnit : int {
  Seconds      = 0,
  Minutes      = 1,
  Hours        = 2,
  Days         = 3
};

enum class ScheduleType : int {
  OneShot      = 0,
  FixedDelta   = 1,
  Weekly       = 2,
  Monthly      = 3,
  Yearly       = 4
};

Fixed schedule mask
delta value between events in ScheduleUnit

Weekly schedule mask
0000001 → Sunday
0000010 → Monday
0000100 → Tuesday
0001000 → Wednesday
0010000 → Thursday
0100000 → Friday
1000000 → Saturday

Monthly schedule mask
Day of the month 1 -31

Yearly schedule mask
| Months since January | Day of the month |
|         0-11         |       1-31       |
|         8 bit        |      8 bit       |
```

pennam commented:
Regarding the msk value i think i can easily add some function to ease creation of scheduler configs. Not sure what to do with frm and to fields... :thinking:

pennam commented:
Hi @aentinger, I still need to fix UnitTests after the last changes, in the meanwhile do you mind having a look to the new example and function to ease Schedule configuration ?

https://github.com/arduino-libraries/ArduinoIoTCloud/pull/284/commits/32292b5ac46c50bd302add21360af671f9f1a0ae also add support for timezone changes from the cloud.

Thanks

pennam commented:
rebased on master and squashed some commits to cleanup history

polldo commented on src/utility/time/TimeService.cpp:
```diff
@@ -91,6 +91,52 @@ unsigned long TimeService::getTime()
+  };
+  static const char month_names[] = "JanFebMarAprMayJunJulAugSepOctNovDec";
+
+  sscanf(timeString.c_str(), "%d %s %d %d:%d:%d", &year, s_month, &day, &hour, &minutes, &seconds);
```
this function is public but there are no checks on the input
for example, if a user executes the following statement, the program will crash
`  Schedule::getTimeFromString("2021 November 01 17:00:00");  `

is this something we care about ?
also, I think that the time format should be documented somewhere

pennam replied:
yes, definitely. I was really unsure about this function, andas you can see here https://github.com/arduino-libraries/ArduinoIoTCloud/pull/284/commits/d317f5631b91f1479089ef519bc6c55b15b64c71 i had tried to use also std::get_time function to do this. Do you think it would be better?

I didn't wont to add much complex checks mostly because i think most people would use the widget to configure the schedule, but for sure i will add some lined to document how to use the function in the example and in the TimeService class

pennam replied:
or maybe jut provide a wrapper around std::get_time provigind the time string and also the information on how is formatted..

polldo replied:
>as you can see here d317f56 i had tried to use also std::get_time function to do this. Do you think it would be better?

yes, using a std function with a standard time format seems better to me 

IMO all other alternatives are also good as long as they are documented (also the one you already implemented)

pennam replied:
fixed with 3cf18b29bd81944820adb793468b2f1b8a8d55e5

polldo commented on src/property/types/CloudSchedule.h:
```diff
@@ -0,0 +1,417 @@
+    }
+
+    static unsigned int createOneShotScheduleConfiguration() {
+      return 0;
```
what should this function do?

pennam replied:
for non repeating schedule the configurazion mask is 0

polldo replied:
oh right!

polldo commented on src/utility/time/TimeService.h:
```diff
@@ -55,8 +59,13 @@ class TimeService
   static bool isTimeValid(unsigned long const time);
 
+private:
+  unsigned long _timezone_offset;
```
the offset can be negative

pennam replied:
Nice catch! thanks

pennam replied:
fixed with cd9a6642f0904630e9d5d747e28439e06b313b20

polldo commented on src/utility/time/TimeService.cpp:
```diff
@@ -129,20 +128,66 @@ unsigned long TimeService::getLocalTime()
+  static const int input_parameters = 6;
+  unsigned long time = 0;
+
+  if(input != nullptr && input.length() == input_length) {
```
```C++
if (input == nullptr || input.length() != input_length) {
       DEBUG_ERROR("ArduinoIoTCloudTCP::%s invalid input length", __FUNCTION__);
       return 0
}
```
I think it'd be more readable this way

pennam replied:
Thanks @polldo, please see 6b9602520d013a1f670abf83e9f08ffe17bb22de

polldo commented on src/utility/time/TimeService.cpp:
```diff
@@ -129,20 +128,66 @@ unsigned long TimeService::getLocalTime()
+      return 0;
+    }
+
+    if(strlen(s_month) > 3) {
```
we may want to check that `strlen(s_month)` is exactly equal to 3 ?

pennam replied:
oh yes -> 6b9602520d013a1f670abf83e9f08ffe17bb22de

polldo commented on src/utility/time/TimeService.cpp:
```diff
@@ -129,20 +128,66 @@ unsigned long TimeService::getLocalTime()
+      return 0;
+    }
+
+    month = (strstr(month_names, s_month) - month_names) / 3;
```
nice pointer arithmetic, but there's a problem if the month is not contained in `month_names` 
I would add a `!= nullptr` check on the result of strstr before using it

pennam replied:
Yes! all credits go to @aentinger :wink: . I've only re-used it badly :smiling_face_with_tear: I've never had to use pointer aritmetic before and i've discovered that null pointer aritmetic is quite an hot topic!! ->https://github.com/arduino-libraries/ArduinoIoTCloud/commit/6b9602520d013a1f670abf83e9f08ffe17bb22de

aentinger replied:
Yeah well, I totally copied that from somewhere else :blush:

aentinger commented on examples/ArduinoIoTCloud-Schedule/ArduinoIoTCloud-Schedule.ino:
```diff
@@ -0,0 +1,229 @@
+  unsigned int activePeriod = MINUTES * 5;
+
+  /* Warning: there is no cross check between until and activePeriod */
+  unsigned int scheduleConfiguration =  Schedule::createOneShotScheduleConfiguration();
```
I was thinking: Why not create a `typedef` to hide the fact that a schedule is nothing more than an `unsigned int`? If at some point in time it becomes a class there's no need change anything, as it already appears like a class due to the `typedef`, i.e.
```C++
typedef unsigned int CloudScheduleType;
```
This could be combined with a `static_assert` to ensure that the cloud type always has the minimum required value, i.e.
```C++
static_assert(sizeof(CloudScheduleType int) >= 4", Error, sizeof(CloudScheduleType) < 4, can't use CloudSchedule on this platform."), 
```

pennam replied:
I'm not sure i get the point, `Schedule` and `CloudSchedule` are already two classes; in the example `scheduleConfiguration` is just one parameter needed to initialize the schedule. That said i think using a typedef for the scheduleConfiguration and the other input parameters is really a good idea so i've created a commit you can check. https://github.com/pennam/ArduinoIoTCloud/commit/2b63f01f2e41c2ad9d75729087eb95bbbf0960e9

aentinger replied:
No, you've gotten me perfectly. :+1: for `ScheduleTimeType` and `ScheduleConfigurationType`.

pennam replied:
-> see https://github.com/arduino-libraries/ArduinoIoTCloud/pull/284/commits/210edd275792da456221fb8d413a884cbcb624be

aentinger commented on src/ArduinoIoTCloud.h:
```diff
@@ -145,7 +147,9 @@ class ArduinoIoTCloudClass
     ConnectionHandler * _connection = nullptr;
     PropertyContainer _property_container;
-    TimeService _time_service;
+    TimeService * _time_service = ArduinoIoTCloudTimeService();
```
Is there any specific reason why this is now a pointer? It looks like you want to employ the `Singleton` design pattern. Why? Usually (except for very few use cases) `Singleton`s are to be avoided.

Also please do all initialisation within the constructor, or all initialisation within the class body, but not mixing between those two options.

aentinger replied:
I see. Still one small improvement would be to operate with references `&` instead of pointers `*`, then you never need to worry that a pointer can be zero, i.e.
```C++
TimeService & ArduinoIoTCloudTimeService() {
  static TimeService _timeService_instance;
  return _timeService_instance;
}
```

aentinger replied:
Related: Why do you have `ArduinoIoTCloudTimeService` twice in code (Once in production code, once in test code). Would it not be sufficient to have it defined only once? I do not see MCU-specific code in that statement that would not compile for host.

pennam replied:
No MCU-specific code, but adding TimeService.cpp to the test system will require some extra work because then we will need also to stub/fake a bit on NTPUtils and maybe also RTC. So at the end i've decided to duplicate the function to make the build system happy and run the tests.

The only property that is actually using TimeService is CloudSchedule so i don't think adding the whole class would be really beneficial for the test coverage since CloudSchedule is implementing its own getTime() fake function in test_CluodSchedule.cpp file.

pennam replied:
> operate with references & instead of pointers *

please see https://github.com/arduino-libraries/ArduinoIoTCloud/pull/284/commits/920e920a7f4125e6272b8c6ba814301eb9f0442f

aentinger commented on extras/test/src/util/PropertyTestUtil.cpp:
```diff
@@ -16,3 +18,13 @@ unsigned long getTime()
+  return getTime();
+}
+
+TimeService* ArduinoIoTCloudTimeService() {
```
This looks somewhat like you want to use a `Singleton` (also see comment below). In general I'd advise to avoid singletons but if you must have them then please follow this pattern:
```C++
/* Singleton.h */
class Singleton
{
public:
  Singleton(Singleton &) = delete;
  void operator = (Singleton &) = delete;

  static Singleton & instance();
  static void destroy();
private:
  static Singleton * _p_instance;
};
/* Singleton.cpp */
Singleton * Singleton::_p_instance{nullptr};
Singleton & Singleton::instance()
{
  if (!_p_instance) {
    _p_instance = new Singleton();
  }
  return *_p_instance;
}

void Singleton::destroy()
{
  delete _p_instance;
  _p_instance = nullptr;
}
```

pennam replied:
implemented with the pattern above using references -> see https://github.com/arduino-libraries/ArduinoIoTCloud/commit/920e920a7f4125e6272b8c6ba814301eb9f0442f

aentinger reviewed (commented):
I love how this PR comes together :heart_eyes:, however, since I'm a nitpicker I still found a couple of suggestions for you :wink:

pennam commented:
>  since I'm a nitpicker I still found a couple of suggestions for you wink

That's the reason why i continue bothering you with my review requests! Thanks!

aentinger commented:
I still suggest to [initialise member variables](https://github.com/arduino-libraries/ArduinoIoTCloud/pull/284/files#diff-2a08667a46c0a45cd028e94148001c11fd307f4d237de0a2f6166c8873777362R151-R152) within the constructor (to stay consistent with the current pattern), but apart from this, this is done imho.

aentinger reviewed (approved):
Love it :+1:

pennam commented:
we need to upgrade that ESP core! :smiling_face_with_tear:

pennam commented:
rebased and squashed some commits to cleanup history before merging.
